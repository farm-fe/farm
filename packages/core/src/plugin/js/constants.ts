export const VITE_DEFAULT_ASSETS = [
  // images
  'apng',
  'png',
  'jpe?g',
  'jfif',
  'pjpeg',
  'pjp',
  'gif',
  'svg',
  'ico',
  'webp',
  'avif',

  // media
  'mp4',
  'webm',
  'ogg',
  'mp3',
  'wav',
  'flac',
  'aac',
  'opus',
  'mov',
  'm4a',
  'vtt',

  // fonts
  'woff2?',
  'eot',
  'ttf',
  'otf',

  // other
  'webmanifest',
  'pdf',
  'txt'
];
// the name of the virtual module internal the adapter
export const VITE_ADAPTER_VIRTUAL_MODULE = 'vite-adapter-virtual:';
