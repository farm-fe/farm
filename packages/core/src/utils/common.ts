import readline from 'node:readline';

export function clearScreen() {
  const repeatCount = process.stdout.rows - 2;
  const blank = repeatCount > 0 ? '\n'.repeat(repeatCount) : '';
  console.log(blank);
  readline.cursorTo(process.stdout, 0, 0);
  readline.clearScreenDown(process.stdout);
}

export function isObject(value: unknown): value is Record<string, unknown> {
  return Object.prototype.toString.call(value) === '[object Object]';
}

export function isEmptyObject<T extends object>(obj: T): boolean {
  return Reflect.ownKeys(obj).length === 0;
}

// Taken from https://github.com/sindresorhus/slash/blob/main/index.js (MIT)
export function slash(path: string) {
  const isExtendedLengthPath = /^\\\\\?\\/.test(path);
  // eslint-disable-next-line no-control-regex
  const hasNonAscii = /[^\u0000-\u0080]+/.test(path);

  if (isExtendedLengthPath || hasNonAscii) {
    return path;
  }

  return path.replace(/\\/g, '/');
}
