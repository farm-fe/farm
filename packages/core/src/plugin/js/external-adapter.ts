import { CUSTOM_KEYS } from '../../config/constants.js';
import {
  mergeCustomExternal,
  partialExternal
} from '../../config/normalize-config/normalize-external.js';
import { UserConfig } from '../../config/types.js';
import { isArray } from '../../utils/share.js';
import { JsPlugin } from '../type.js';

/**
 * avoid add new external in config hook
 */
export function externalAdapter(): JsPlugin {
  return {
    name: 'farm:external-adapter',

    priority: -Infinity,

    config(config: UserConfig): UserConfig | Promise<UserConfig> {
      if (
        config?.compilation?.external &&
        isArray(config.compilation.external)
      ) {
        let [stringExternal, recordExternal] = mergeCustomExternal(
          config?.compilation,
          partialExternal(config.compilation.external)
        );

        return {
          compilation: {
            external: stringExternal,
            custom: {
              [CUSTOM_KEYS.external_record]: JSON.stringify(recordExternal)
            }
          }
        };
      }
    }
  };
}
