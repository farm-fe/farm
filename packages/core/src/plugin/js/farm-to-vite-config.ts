import type { UserConfig as ViteUserConfig } from 'vite';
import type { UserConfig } from '../../config/types.js';
import {
  deleteUndefinedPropertyDeeply,
  throwIncompatibleError
} from './utils.js';
import merge from '../../utils/merge.js';
import { Logger } from '../../index.js';
import { VITE_DEFAULT_ASSETS } from './constants.js';

export function farmUserConfigToViteConfig(config: UserConfig): ViteUserConfig {
  const vitePlugins = config.vitePlugins.filter(Boolean).map((plugin) => {
    if (typeof plugin === 'function') {
      return plugin().vitePlugin;
    } else {
      return plugin;
    }
  });

  let sourcemap = true;

  if (config.compilation?.sourcemap !== undefined) {
    sourcemap = Boolean(config.compilation?.sourcemap);
  }

  const viteConfig: ViteUserConfig = {
    root: config.root,
    base: config.compilation?.output?.publicPath ?? '/',
    publicDir: config.publicDir ?? 'public',
    mode: config.compilation?.mode,
    define: config.compilation?.define,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore ignore this error
    command: config.compilation?.mode === 'production' ? 'build' : 'serve',
    resolve: {
      alias: config.compilation?.resolve?.alias,
      extensions: config.compilation?.resolve?.extensions,
      mainFields: config.compilation?.resolve?.mainFields,
      conditions: config.compilation?.resolve?.conditions,
      preserveSymlinks: config.compilation?.resolve?.symlinks === false
    },
    plugins: vitePlugins,
    server: {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore ignore error
      hmr: config.server?.hmr,
      port: config.server?.port,
      host: config.server?.host,
      strictPort: config.server?.strictPort,
      https: config.server?.https,
      proxy: config.server?.proxy as any,
      open: config.server?.open,
      watch:
        typeof config.server?.hmr === 'object'
          ? config.server.hmr?.watchOptions ?? {}
          : {}
      // other options are not supported in farm
    },
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore ignore this error
    isProduction: config.compilation?.mode === 'production',
    css: config.compilation?.css?._viteCssOptions ?? {},
    build: {
      outDir: config.compilation?.output?.path,
      sourcemap,
      minify:
        config.compilation?.minify !== undefined
          ? Boolean(config.compilation?.minify)
          : undefined,
      cssMinify:
        config.compilation?.minify !== undefined
          ? Boolean(config.compilation?.minify)
          : undefined,
      ssr: config.compilation?.output?.targetEnv === 'node',
      rollupOptions: {
        output: {
          assetFileNames: config.compilation?.output?.assetsFilename,
          entryFileNames: config.compilation?.output?.entryFilename,
          chunkFileNames: config.compilation?.output?.filename
        }
      }
      // other options are not supported in farm
    },
    // TODO make it configurable
    cacheDir: 'node_modules/.farm/cache',
    envDir: config.envDir,
    assetsInclude: [
      ...VITE_DEFAULT_ASSETS,
      ...(config.compilation?.assets?.include ?? [])
    ]
  };

  return viteConfig;
}

export function proxyViteConfig(
  viteConfig: ViteUserConfig,
  pluginName: string,
  logger: Logger
): ViteUserConfig {
  return new Proxy(viteConfig, {
    get(target, key) {
      if (typeof key !== 'string') {
        return target[key as unknown as keyof typeof target];
      }

      if (
        key === 'then' ||
        key === 'length' ||
        key === 'constructor' ||
        key === 'prototype'
      ) {
        return (target as Record<string, any>)[key];
      }

      const allowedKeys = [
        'root',
        'base',
        'publicDir',
        'mode',
        'define',
        'command',
        'resolve',
        'plugins',
        'server',
        'isProduction',
        'css',
        'build',
        'logger',
        'cacheDir',
        'envDir',
        'assetsInclude',
        // these fields are always undefined in farm
        // they are only used for compatibility
        'legacy',
        'optimizeDeps',
        'ssr',
        'logLevel',
        'experimental',
        'test',
        'clearScreen',
        'customLogger'
      ];

      if (allowedKeys.includes(String(key))) {
        if (key === 'resolve') {
          const allowedResolveKeys = [
            'alias',
            'extensions',
            'mainFields',
            'conditions',
            'preserveSymlinks',
            // farm do not set any thing for dedupe, it should always be undefined
            'dedupe'
          ];

          return new Proxy(target.resolve || {}, {
            get(resolveTarget, resolveKey) {
              if (typeof resolveKey !== 'string') {
                return target[resolveKey as unknown as keyof typeof target];
              }

              if (allowedResolveKeys.includes(String(resolveKey))) {
                return resolveTarget[resolveKey as keyof typeof resolveTarget];
              }

              if (
                resolveKey === 'then' ||
                resolveKey === 'length' ||
                resolveKey === 'constructor' ||
                resolveKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.resolve',
                allowedResolveKeys,
                resolveKey
              );
            }
          });
        } else if (key === 'server') {
          const allowedServerKeys = [
            'hmr',
            'port',
            'host',
            'strictPort',
            'https',
            'proxy',
            'open',
            'origin',
            'watch'
          ];

          return new Proxy(target.server || {}, {
            get(serverTarget, serverKey) {
              if (typeof serverKey !== 'string') {
                return target[serverKey as unknown as keyof typeof target];
              }

              if (allowedServerKeys.includes(String(serverKey))) {
                return serverTarget[serverKey as keyof typeof serverTarget];
              }

              if (
                serverKey === 'then' ||
                serverKey === 'length' ||
                serverKey === 'constructor' ||
                serverKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.server',
                allowedServerKeys,
                serverKey
              );
            }
          });
        } else if (key === 'css') {
          const allowedCssKeys = [
            'devSourcemap',
            'transformer',
            'modules',
            'postcss',
            'preprocessorOptions'
          ];

          return new Proxy(target.css || {}, {
            get(cssTarget, cssKey) {
              if (typeof cssKey !== 'string') {
                return target[cssKey as unknown as keyof typeof target];
              }

              if (allowedCssKeys.includes(String(cssKey))) {
                return cssTarget[cssKey as keyof typeof cssTarget];
              }

              if (
                cssKey === 'then' ||
                cssKey === 'length' ||
                cssKey === 'constructor' ||
                cssKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.css',
                allowedCssKeys,
                cssKey
              );
            }
          });
        } else if (key === 'build') {
          const allowedBuildKeys = [
            'outDir',
            'sourcemap',
            'minify',
            'cssMinify',
            'ssr',
            'watch',
            'rollupOptions',
            'assetsDir'
          ];

          return new Proxy(target.build || {}, {
            get(buildTarget, buildKey) {
              if (typeof buildKey !== 'string') {
                return target[buildKey as unknown as keyof typeof target];
              }

              if (allowedBuildKeys.includes(String(buildKey))) {
                return buildTarget[buildKey as keyof typeof buildTarget];
              }
              if (
                buildKey === 'then' ||
                buildKey === 'length' ||
                buildKey === 'constructor' ||
                buildKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.build',
                allowedBuildKeys,
                buildKey
              );
            }
          });
        } else if (key === 'optimizeDeps') {
          return new Proxy(target.optimizeDeps || {}, {
            get(_, optimizeDepsKey) {
              logger.warnOnce(
                `[vite-plugin] ${pluginName}: config "optimizeDeps" is not needed in farm, all of its options will be ignored. Current ignored option is: "${String(
                  optimizeDepsKey
                )}"`
              );

              if (optimizeDepsKey === 'esbuildOptions') {
                return {};
              }
              return undefined;
            }
          });
        } else if (key === 'logger') {
          return logger;
        } else if (key === 'assetsInclude') {
          return (filename: string) => {
            return (
              (viteConfig.assetsInclude as string[])?.some((r) => {
                return new RegExp(r).test(filename);
              }) ?? false
            );
          };
        }

        return target[key as keyof typeof target];
      }

      throw throwIncompatibleError(pluginName, 'viteConfig', allowedKeys, key);
    }
  });
}

export function viteConfigToFarmConfig(
  config: ViteUserConfig,
  origFarmConfig: UserConfig,
  _pluginName: string
): UserConfig {
  const farmConfig: UserConfig = {
    compilation: {}
  };

  if (config.root) {
    farmConfig.root = config.root;
  }
  if (config?.css) {
    farmConfig.compilation.css ??= {};
    farmConfig.compilation.css._viteCssOptions = config.css;
  }

  if (config.base) {
    farmConfig.compilation.output ??= {};
    farmConfig.compilation.output.publicPath = config.base;
  }
  if (config.publicDir) {
    farmConfig.publicDir = config.publicDir;
  }
  if (config.mode === 'development' || config.mode === 'production') {
    farmConfig.compilation.mode = config.mode;
  }
  if (config.define) {
    farmConfig.compilation.define = config.define;
  }
  if (config.resolve) {
    farmConfig.compilation.resolve ??= {};

    if (config.resolve.alias) {
      if (!Array.isArray(config.resolve.alias)) {
        farmConfig.compilation.resolve.alias = config.resolve.alias as Record<
          string,
          any
        >;
      } else {
        if (!farmConfig.compilation.resolve.alias) {
          farmConfig.compilation.resolve.alias = {};
        }

        const farmRegexPrefix = '$__farm_regex:';

        for (const { find, replacement } of config.resolve.alias) {
          if (find instanceof RegExp) {
            const key = farmRegexPrefix + find.source;
            farmConfig.compilation.resolve.alias[key] = replacement;
          } else {
            farmConfig.compilation.resolve.alias[find] = replacement;
          }
        }
      }
    }

    farmConfig.compilation.resolve.extensions = config.resolve.extensions;
    farmConfig.compilation.resolve.mainFields = config.resolve.mainFields;
    farmConfig.compilation.resolve.conditions = config.resolve.conditions;
    farmConfig.compilation.resolve.symlinks =
      config.resolve.preserveSymlinks != true;
  }

  if (config.server) {
    farmConfig.server ??= {};
    farmConfig.server.hmr = config.server.hmr;
    farmConfig.server.port = config.server.port;

    if (config.server.watch) {
      if (
        farmConfig.server?.hmr === true ||
        farmConfig.server?.hmr === undefined
      ) {
        farmConfig.server.hmr = {
          ...(typeof origFarmConfig?.server?.hmr === 'object'
            ? origFarmConfig.server.hmr
            : {}),
          watchOptions: config.server.watch
        };
      }
    }

    if (typeof config.server.host === 'string') {
      farmConfig.server.host = config.server.host;
    }

    farmConfig.server.strictPort = config.server.strictPort;
    farmConfig.server.https =
      typeof config.server.https === 'boolean'
        ? undefined
        : config.server.https;
    farmConfig.server.proxy = config.server.proxy as any;
    farmConfig.server.open = Boolean(config.server.open);
  }

  if (config.build) {
    farmConfig.compilation.output ??= {};
    farmConfig.compilation.output.path = config.build.outDir;

    if (
      config.build?.sourcemap !== undefined &&
      origFarmConfig.compilation?.sourcemap === undefined
    ) {
      farmConfig.compilation.sourcemap = Boolean(config.build.sourcemap);
    }

    if (config.build?.minify !== undefined) {
      farmConfig.compilation.minify = Boolean(config.build.minify);
    }

    if (
      config.build.ssr !== undefined &&
      origFarmConfig.compilation?.lazyCompilation === undefined
    ) {
      farmConfig.compilation.lazyCompilation = !config.build.ssr;
    }

    if (config.build.rollupOptions?.output !== undefined) {
      if (!Array.isArray(config.build.rollupOptions.output)) {
        const keys = ['assetFileNames', 'entryFilename', 'filename'];

        for (const k of keys) {
          /* eslint-disable @typescript-eslint/ban-ts-comment */
          // @ts-ignore type is correct
          farmConfig.compilation.output[k] =
            // @ts-ignore type is correct
            config.build.rollupOptions.output[k];
        }
      }
    }
  }

  deleteUndefinedPropertyDeeply(farmConfig);

  return merge({}, origFarmConfig, farmConfig);
}
