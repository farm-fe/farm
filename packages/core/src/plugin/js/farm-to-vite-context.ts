import type { PluginContext } from 'rollup';
import type { UserConfig } from '../../config/types.js';
import type { CompilationContext } from '../type.js';
import { relative } from 'path';
import { DefaultLogger } from '../../utils/logger.js';

const contextCache = new Map<string, PluginContext>();

export function farmContextToViteContext(
  farmContext: CompilationContext,
  currentHandlingFile?: string,
  pluginName?: string,
  hookName?: string,
  config?: UserConfig
): PluginContext {
  const cacheKey = pluginName + hookName;
  if (contextCache.has(cacheKey)) {
    return contextCache.get(cacheKey) as PluginContext;
  }
  const logger = new DefaultLogger();

  const log = (message: any) => {
    if (typeof message === 'function') {
      message = message();
    }

    console.log(message);
  };

  const cacheError = () => {
    throw new Error(
      `Vite plugin ${pluginName} is not compatible with Farm for now. Because cache(called by hook ${pluginName}.${hookName}) is not supported in Farm`
    );
  };

  const viteContext: PluginContext = {
    addWatchFile: (id) => {
      if (!currentHandlingFile) {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because addWatchFile(called by hook ${pluginName}.${hookName}) can only be called in load hook or transform hook in Farm.`
        );
      }
      farmContext.addWatchFile(currentHandlingFile, id);
    },
    debug: log,
    emitFile: (params) => {
      if (params.type === 'asset') {
        let content: number[] = [];

        if (typeof params.source === 'string') {
          content = [...Buffer.from(params.source)];
        } else {
          content = [...params.source];
        }

        farmContext.emitFile({
          resolvedPath: currentHandlingFile ?? 'vite-plugin-adapter',
          name: params.fileName ?? params.name,
          content,
          resourceType: 'asset'
        });

        return 'vite-plugin-adapter-unsupported-reference-id';
      } else {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because emitFile(called by hook ${pluginName}.${hookName}) can only emit asset in Farm.`
        );
      }
    },
    error: (message): never => {
      if (typeof message === 'object') {
        farmContext.error(JSON.stringify(message));
      } else {
        farmContext.error(message);
      }

      return undefined as unknown as never;
    },
    getFileName: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getFileName(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleIds: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleIds(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleInfo: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleInfo(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getWatchFiles: () => {
      return farmContext.getWatchFiles();
    },
    info: log,
    load: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because load(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    meta: {
      rollupVersion: '3.29.4',
      watchMode: config.compilation.mode !== 'production'
    },
    parse: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because parse(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    resolve: async (source, importer, options) => {
      if (options.custom.caller === `${pluginName}.${hookName}`) {
        return null;
      }

      const farmResolveResult = await farmContext.resolve(
        {
          source,
          importer: {
            relativePath: relative(config.compilation.root, importer),
            queryString: importer.split('?')[1] ?? ''
          },
          kind: options.isEntry ? 'entry' : 'import'
        },
        {
          meta: {},
          caller: `${pluginName}.${hookName}`
        }
      );

      if (farmResolveResult) {
        return {
          id: farmResolveResult.resolvedPath,
          external: farmResolveResult.external,
          resolvedBy: 'vite-plugin-adapter-farm-resolve',
          moduleSideEffects: farmResolveResult.sideEffects,
          meta: {
            ...farmResolveResult.meta,
            caller: `${pluginName}.${hookName}`
          },
          // TODO these 2 options are not supported in farm
          assertions: {},
          syntheticNamedExports: false
        };
      }

      return null;
    },
    setAssetSource(assetReferenceId, source) {
      this.emitFile({
        type: 'asset',
        source,
        name: assetReferenceId
      });
    },
    warn: (message) => {
      if (typeof message === 'object') {
        farmContext.warn(JSON.stringify(message));
      } else if (typeof message === 'function') {
        farmContext.warn(JSON.stringify(message()));
      } else {
        farmContext.warn(message);
      }
    },
    cache: {
      set: cacheError,
      get: cacheError,
      delete: cacheError,
      has: cacheError
    },
    moduleIds: new Set<string>()[Symbol.iterator](),
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore Vite specific property
    getCombinedSourcemap() {
      logger.warn(
        '`vite-plugin-adapter`: getCombinedSourcemap is not supported in Farm for now. It will always return undefined.'
      );
      return undefined;
    }
  };

  contextCache.set(cacheKey, viteContext);

  return viteContext;
}
