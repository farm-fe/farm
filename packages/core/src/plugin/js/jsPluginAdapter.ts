import { toArray } from '../../utils/index.js';

export function adaptorVitePlugin<UserOptions = Record<string, never>>(
  factory: any
) {
  return (userOptions?: UserOptions) => {
    const meta: any = {
      framework: 'vite'
    };
    const rawPlugins = toArray(factory(userOptions!, meta));

    const plugins = rawPlugins.map((rawPlugin: any) => {
      const plugin = transformFarmPlugin(rawPlugin);

      return plugin;
    });

    return plugins.length === 1 ? plugins[0] : plugins;
  };
}

function transformFarmPlugin(plugin: any) {
  console.log(plugin);

  return {};
}
