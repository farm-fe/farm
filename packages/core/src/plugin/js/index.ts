import { fromZodError } from 'zod-validation-error';
import { CompilationMode } from '../../config/env.js';
import { type JsPlugin, type UserConfig } from '../../index.js';
import merge from '../../utils/merge.js';
import { resolveAsyncPlugins } from '../index.js';

import { cssPluginUnwrap, cssPluginWrap } from './adapter-plugins/css.js';
import { defaultLoadPlugin } from './adapter-plugins/default-load.js';
import {
  PluginSchemaRegistry,
  createAugmentResourceHashSchema,
  createBuildEndSchema,
  createBuildStartSchema,
  createConfigResolvedSchema,
  createConfigSchema,
  createConfigureCompilerSchema,
  createConfigureServerSchema,
  createFinalizeResourcesSchema,
  createFinishSchema,
  createLoadSchema,
  createNameSchema,
  createPluginCacheLoadedSchema,
  createPrioritySchema,
  createRenderResourcePotSchema,
  createRenderStartSchema,
  createResolveSchema,
  createTransformHtmlSchema,
  createTransformSchema,
  createUpdateFinishedSchema,
  createUpdateModulesSchema,
  createWritePluginCacheSchema,
  createWriteResourcesSchema
} from './js-plugin-schema.js';

import { normalizeDevServerConfig } from '../../config/resolve-server.js';
import { DEFAULT_FILTERS, normalizeFilterPath } from './utils.js';
import { VitePluginAdapter } from './vite-plugin-adapter.js';

export { VitePluginAdapter } from './vite-plugin-adapter.js';
export * from './js-plugin-schema.js';
export * from './utils.js';

type VitePluginType = object | (() => { vitePlugin: any; filters: string[] });
type VitePluginsType = VitePluginType[];

export async function handleVitePlugins(
  vitePlugins: VitePluginsType,
  userConfig: UserConfig,
  mode: CompilationMode
): Promise<JsPlugin[]> {
  const jsPlugins: JsPlugin[] = [];
  const filtersUnion = new Set<string>();

  if (vitePlugins.length) {
    userConfig = merge({}, userConfig, {
      compilation: userConfig.compilation,
      server: normalizeDevServerConfig(userConfig)
    });
  }
  const flatVitePlugins = await resolveAsyncPlugins(vitePlugins);

  for (const vitePluginObj of flatVitePlugins) {
    let vitePlugin = vitePluginObj,
      filters = DEFAULT_FILTERS;

    if (typeof vitePluginObj === 'function') {
      const { vitePlugin: plugin, filters: f } = vitePluginObj();
      vitePlugin = plugin;
      filters = f;
    }
    filters?.forEach((filter) => filtersUnion.add(filter));
    processVitePlugin(vitePlugin, userConfig, filters, jsPlugins, mode);
  }

  // if vitePlugins is not empty, append a load plugin to load file
  // this plugin is only for compatibility
  if (vitePlugins.length) {
    jsPlugins.push(
      defaultLoadPlugin({
        filtersUnion,
        userConfig
      })
    );
    jsPlugins.unshift(cssPluginWrap({ filtersUnion }));
    jsPlugins.push(cssPluginUnwrap({ filtersUnion }));
  }

  return jsPlugins;
}

export function processVitePlugin(
  vitePlugin: VitePluginType,
  userConfig: UserConfig,
  filters: string[],
  jsPlugins: JsPlugin[],
  mode: CompilationMode
) {
  const processPlugin = (plugin: any) => {
    const vitePluginAdapter = new VitePluginAdapter(
      plugin as any,
      userConfig,
      filters,
      mode
    );
    // @ts-ignore
    convertPluginVite(vitePluginAdapter);
    jsPlugins.push(vitePluginAdapter);
  };

  if (Array.isArray(vitePlugin)) {
    vitePlugin.forEach((plugin) => processPlugin(plugin));
  } else {
    processPlugin(vitePlugin);
  }
}

const schemaRegistry = new PluginSchemaRegistry();

schemaRegistry
  .register('name', createNameSchema)
  .register('priority', createPrioritySchema)
  .register('configureServer', createConfigureServerSchema)
  .register('configureCompiler', createConfigureCompilerSchema)
  .register('config', createConfigSchema)
  .register('configResolved', createConfigResolvedSchema)
  .register('buildStart', createBuildStartSchema)
  .register('resolve', createResolveSchema)
  .register('load', createLoadSchema)
  .register('transform', createTransformSchema)
  .register('buildEnd', createBuildEndSchema)
  .register('renderStart', createRenderStartSchema)
  .register('processRenderedResourcePot', createRenderResourcePotSchema)
  .register('augmentResourceHash', createAugmentResourceHashSchema)
  .register('finalizeResources', createFinalizeResourcesSchema)
  .register('transformHtml', createTransformHtmlSchema)
  .register('writeResources', createWriteResourcesSchema)
  .register('pluginCacheLoaded', createPluginCacheLoadedSchema)
  .register('writePluginCache', createWritePluginCacheSchema)
  .register('finish', createFinishSchema)
  .register('updateFinished', createUpdateFinishedSchema)
  .register('updateModules', createUpdateModulesSchema);

export function convertPlugin(plugin: JsPlugin) {
  try {
    // TODO process module hook and freeze module hook schema
    const pluginSchema = schemaRegistry.createPluginSchema(plugin?.name);
    return pluginSchema.parse(plugin);
  } catch (err) {
    const validationError = fromZodError(err, {
      prefix: 'Failed to verify js plugin schema'
    });
    const pluginName = plugin?.name || 'undefined';
    throw new Error(
      `${validationError.toString()}. \n Please check '${pluginName}' plugin passes these attributes correctly.`
    );
  }
}

export function convertPluginVite(plugin: JsPlugin): void {
  if (
    plugin.transform &&
    !plugin.transform.filters?.moduleTypes &&
    !plugin.transform.filters?.resolvedPaths
  ) {
    throw new Error(
      `transform hook of plugin ${plugin.name} must have at least one filter(like moduleTypes or resolvedPaths)`
    );
  }
  if (plugin.transform) {
    if (!plugin.transform.filters.moduleTypes) {
      plugin.transform.filters.moduleTypes = [];
    } else if (!plugin.transform.filters.resolvedPaths) {
      plugin.transform.filters.resolvedPaths = [];
    }
  }

  if (plugin.processModule) {
    plugin.processModule.filters ??= {};
    plugin.processModule.filters.moduleTypes ??= [];
    plugin.processModule.filters.resolvedPaths ??= [];
  }

  if (plugin.processRenderedResourcePot) {
    plugin.processRenderedResourcePot.filters ??= {};

    if (
      !plugin.processRenderedResourcePot?.filters?.moduleIds &&
      !plugin.processRenderedResourcePot?.filters?.resourcePotTypes
    ) {
      throw new Error(
        `processRenderedResourcePot hook of plugin ${plugin.name} must have at least one filter(like moduleIds or resourcePotTypes)`
      );
    }

    if (!plugin.processRenderedResourcePot.filters?.resourcePotTypes) {
      plugin.processRenderedResourcePot.filters.resourcePotTypes = [];
    } else if (!plugin.processRenderedResourcePot.filters?.moduleIds) {
      plugin.processRenderedResourcePot.filters.moduleIds = [];
    }
  }

  if (plugin.augmentResourcePotHash) {
    plugin.augmentResourcePotHash.filters ??= {};

    if (
      !plugin.augmentResourcePotHash?.filters?.moduleIds &&
      !plugin.augmentResourcePotHash?.filters?.resourcePotTypes
    ) {
      throw new Error(
        `augmentResourcePotHash hook of plugin ${plugin.name} must have at least one filter(like moduleIds or resourcePotTypes)`
      );
    }

    if (!plugin.augmentResourcePotHash.filters?.resourcePotTypes) {
      plugin.augmentResourcePotHash.filters.resourcePotTypes = [];
    } else if (!plugin.augmentResourcePotHash.filters?.moduleIds) {
      plugin.augmentResourcePotHash.filters.moduleIds = [];
    }
  }

  if (plugin.resolve?.filters?.importers?.length) {
    plugin.resolve.filters.importers =
      plugin.resolve.filters.importers.map(normalizeFilterPath);
  }

  if (plugin.load?.filters?.resolvedPaths?.length) {
    plugin.load.filters.resolvedPaths =
      plugin.load.filters.resolvedPaths.map(normalizeFilterPath);
  }

  if (plugin.transform?.filters?.resolvedPaths?.length) {
    plugin.transform.filters.resolvedPaths =
      plugin.transform.filters.resolvedPaths.map(normalizeFilterPath);
  }
  if (plugin.augmentResourcePotHash?.filters?.moduleIds) {
    plugin.augmentResourcePotHash.filters.moduleIds =
      plugin.augmentResourcePotHash.filters.moduleIds.map(normalizeFilterPath);
  }

  if (plugin.processRenderedResourcePot?.filters?.moduleIds) {
    plugin.processRenderedResourcePot.filters.moduleIds =
      plugin.processRenderedResourcePot.filters.moduleIds.map(
        normalizeFilterPath
      );
  }
}
