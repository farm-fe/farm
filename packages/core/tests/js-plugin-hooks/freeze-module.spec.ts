import { expect, test } from 'vitest';
import { getOutputResult } from '../common.js';
import { getCompiler, getOutputFilePath } from './common.js';

test('Js Plugin Execution - freezeModule', async () => {
  const hookName = 'freeze-module';
  const calledHooks: string[] = [];
  const calledModules: string[] = [];
  const compiler = await getCompiler(
    '',
    [
      {
        name: `test-${hookName}`,
        priority: 1000,
        freezeModule: {
          filters: {
            resolvedPaths: ['^index.ts\\?foo=bar$']
            // moduleTypes: ['js']
          },
          executor: async (param) => {
            expect(param.content).toContain(`${hookName}-return-value`);

            expect(param).matchSnapshot();
            calledHooks.push(hookName);
            calledModules.push(param.moduleId);

            return {
              content: param.content.replace(`${hookName}-return-value`, '1')
            };
          }
        }
      }
    ],
    hookName,
    undefined,
    undefined,
    {
      sourcemap: true
    }
  );

  console.log('compile');
  await compiler.compile();
  console.log('compile end');
  compiler.writeResourcesToDisk();

  expect(calledHooks).toEqual([hookName]);
  expect(calledModules).toEqual(['index.ts?foo=bar']);

  const outputFilePath = getOutputFilePath('', hookName);
  const result = await getOutputResult(outputFilePath);
  expect(result.default).toBe('1');
});
