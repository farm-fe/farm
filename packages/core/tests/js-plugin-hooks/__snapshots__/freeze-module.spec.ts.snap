// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Js Plugin Execution - freezeModule 1`] = `
{
  "content": "import Dep from './dep.js';
console.log(Dep);
export default 'freeze-module-return-value';
",
  "moduleId": "index.ts?foo=bar",
  "moduleType": "js",
  "resolvedDeps": [
    [
      "dep.ts",
      [
        {
          "kind": "import",
          "order": 0,
          "source": "./dep.js",
        },
      ],
    ],
  ],
  "sourceMapChain": [
    "{"version":3,"sources":["index.ts_foo_bar"],"names":[],"mappings":"AAAA,OAAO,SAAS,WAAW;AAE3B,QAAQ,GAAG,CAAC;AAEZ,eAAe,6BAA6B"}",
  ],
}
`;
